src/Microsoft.OData.Core/Metadata/EdmLibraryExtensions.cs

```diff
@@ -252,17 +254,129 @@ internal static IEnumerable<IEdmOperation> FilterBoundOperationsWithSameTypeHier
+        }
+
+        /// <summary>
+        /// Filters the operations candidates based on parameters.
+        /// </summary>
+        /// <param name="operations">The operations.</param>
+        /// <param name="bindingType">Binding type for the operation.</param>
+        /// <param name="parameterNameList">The parameters.</param>
+        /// <param name="caseInsensitive">Whether case insensitive.</param>
+        /// <param name="actionItems">outputs the actions found while matching functions</param>
+        /// <returns>Operations filtered by parameter.</returns>
+        internal static IList<IEdmOperation> FilterOperationCandidatesBasedOnParameterList(this IEnumerable<IEdmOperation> operations, IEdmType bindingType, IList<string> parameterNameList, bool caseInsensitive, out IList<IEdmOperation> actionItems)
+        {
+            Debug.Assert(operations != null, "operations");
+            Debug.Assert(parameterNameList != null, "parameters");
+
+            bool hasParameters = parameterNameList.Count > 0;
+            actionItems = null;
+            IList<IEdmOperation> operationCandidates = new List<IEdmOperation>();
+
+            foreach (var operation in operations)
+            {
+                if (bindingType != null)
+                {
+                    operation.EnsureOperationBoundWithBindingParameter();
+                }
+
+                bool suitableCandidate = true;
+
+                // If the number of binding parameters > 0 then we should only keep functions as actions can't have parameters on the uri only in the payload. Filter further by parameters in this case, otherwise don't.
+                // This logic might be a potential bug. 
+                if (hasParameters)
+                {
+                    if (operation.IsAction())
+                    {
+                        if (actionItems == null)
+                        {
+                            actionItems = new List<IEdmOperation>();
+                        }
+
+                        actionItems.Add(operation);
+                        suitableCandidate = false;
+                    }
+
+                    if (!ParametersSatisfyFunction(operation, parameterNameList, caseInsensitive))
+                    {
+                        suitableCandidate = false;
+                    }
+                }
+                else if (bindingType != null)
+                {
+                    // If it is a bound function without required parameters, then the parameter list should only have one parameter and any others as optional.
+                    if (operation.IsFunction())
+                    {
+                        int count = 0;
+                        foreach (var param in operation.Parameters)
+                        {
+                            if (count > 0 && !(param is IEdmOptionalParameter))
+                            {
+                                suitableCandidate = false;

```

Comment:
See if edmlib enforces optional parameters at the end then just look for the first optional parameter.

Reply:
EdmLib does not validate the ordering. Just checking for the first optional parameter fails some of our tests. 

---
In reply to: [370881673](https://github.com/OData/odata.net/pull/1602#discussion_r370881673) [](ancestors = 370881673)